### PR TITLE
Wrapper has the same interface as the wrapped function

### LINKED ISSUE #6: support for missing arguments non-referentially transparent
I noticed that  memoise assumes that the only arguments that can change are those provided by the user as implemented in these two lines of code.

```
 memo_f <- function(...) {
    hash <- digest(list(...))
```

If the function to be memoised  also has default arguments that can change, this won't work

```
g = function(x = rnorm(1))x
g()
[1] -0.2685467
g()
[1] -1.901578
g()
[1] -0.9886473
h = memoise(g)
h()
[1] 0.07904293
 h()
[1] 0.07904293
h()
[1] 0.07904293
h()
[1] 0.07904293
```

My use case for this is when an argument is a filename whose contents can change, but not always,  between invocations. Then memoising the function as is won't work, but if we added an argument with a default value set to `file.info` for that same file, then `file.info` would change every time the file contents are changed, at least, and trigger the computation. I think one way to implement this would be to access the missing arguments with `formals` and selectively evaluate them. A completely different approach would be to allow memoise users to specify how to eval certain arguments for the purpose of hashing, as in

```
memoise(f, custom.eval = list(filename = function(fn) file.info(fn))
```

This is less intrusive because it doesn't require  a change in the f function signature,  but makes `memoise` more complicated. In the filename example we'd have

```
f = function(fname, timestamp = file.info(fname) ....
h = memoise(f) #will find timestamp through formals
```

as opposed to

```
f = function(fname) ...
h = memoise(f, custom.eval = list(fname = file.info(fname))
```

Makes sense? Would a pull request towards either of this be considered?

### LINKED ISSUE #12: memoise breaks autocompletion in RStudio
This is just a small inconvenience.
It appears that applying `memoise()` to a function breaks the argument autocompletion for such a `memoise`d function –– at least in RStudio.
Not sure on whose end that's a problem.

### PR DESCRIPTION
**Pull request by krlmlr**
Also, the wrapped function is referred to by `memoised_function`, which is clearer than `f`.

Cc @maxheld83 and @piccolbo.

Fixes #12.

Fixes #6.

Closes #7 (=supersedes it).

Closes #2 (=includes it).

Closes #13 (=includes it).

Example:

```
> memoize(runif)
function (n, min = 0, max = 1) 
{
    hash <- digest(list(n, min, max))
    if (cache$has_key(hash)) {
        res <- cache$get(hash)
    }
    else {
        res <- withVisible(memoised_function(n = n, min = min, 
            max = max))
        cache$set(hash, res)
    }
    if (res$visible) {
        res$value
    }
    else {
        invisible(res$value)
    }
}
<environment: 0x3c32040>
attr(,"memoised")
[1] TRUE
> memoize(paste)
function (..., sep = " ", collapse = NULL) 
{
    hash <- digest(list(..., sep, collapse))
    if (cache$has_key(hash)) {
        res <- cache$get(hash)
    }
    else {
        res <- withVisible(memoised_function(... = ..., sep = sep, 
            collapse = collapse))
        cache$set(hash, res)
    }
    if (res$visible) {
        res$value
    }
    else {
        invisible(res$value)
    }
}
<environment: 0x3d687f0>
attr(,"memoised")
[1] TRUE
```

### CODE REVIEW

**Comment by krlmlr** [comment]
Now complete, including revdep-check. If the memoized function is not yet defined during memoization, its formals cannot be queried, and the old implementation is used (with a warning). Ironically, the package that pointed me at that issue -- lubridate -- doesn't use memoise anymore. (It still fails, but this is due to locale settings on my machine.)

The [individual merges](https://github.com/krlmlr/memoise/pulls?q=is%3Apr+is%3Aclosed+no%3Alabel+sort%3Aupdated-asc) (excluding trivial ones, oldest first) provide additional detail.

File: R/memoise.r [hunk 1]
```diff
@@ -35,64 +35,132 @@
...
-#' # (If you're not sure whether you already memoised something,  
+#' # (If you're not sure whether you already memoised something,
 #' #  use is.memoised() to check.)
 #' memA(2)
 #' memA <- memoise(a)
 #' memA(2)
-memoise <- memoize <- function(f) {
+memoise <- memoize <- function(f, envir = parent.frame()) {
+  # We must not even try evaluating f -- once we start, there's no way back
+  if (inherits(try(eval.parent(substitute(f)), silent = TRUE), "try-error")) {
```
Comment (hadley) [inline R/memoise.r]: Is there a reason you don't want to use exists here?

File: R/memoise.r [hunk 2]
```diff
@@ -35,64 +35,132 @@
...
+            "Define the memoised function before memoising to avoid this warning.")
+    memoise_old(f)
+  } else {
+    memoise_new(f, envir)
+  }
+}
+
+#' @importFrom stats setNames
+memoise_new <- function(f, envir) {
+  f_formals <- formals(f)
```
Comment (hadley) [inline R/memoise.r]: I think you also need a test for primitive functions (e.g. `+`), which don't have formals. They should use `memoise_old()` without a warning.

File: R/memoise.r [hunk 3]
```diff
@@ -35,64 +35,132 @@
...
+  f_formal_names <- names(f_formals)
+  f_formal_name_list <- lapply(f_formal_names, as.name)
+
+  # list(...)
+  list_call <- as.call(c(list(as.name("list")), f_formal_name_list))
```
Comment (hadley) [inline R/memoise.r]: I prefer `quote(list)` to `as.name("list")`

File: R/memoise.r [hunk 4]
```diff
@@ -35,64 +35,132 @@
...
+
+  # memoised_function(...)
+  init_call_args <- setNames(f_formal_name_list, f_formal_names)
+  init_call <- as.call(c(as.name("memoised_function"), init_call_args))
+
   cache <- new_cache()
-  
+
+  memo_f <- eval(
```
Comment (hadley) [inline R/memoise.r]: Could you use this instead?

```R
make_function <- function(args, body, env = parent.frame()) {
  args <- as.pairlist(args)

  eval(call("function", args, body), env)
}
```

File: R/memoise.r [hunk 5]
```diff
@@ -35,64 +35,132 @@
...
-      cache$get(hash)
+      res <- cache$get(hash)
     } else {
-      res <- f(...)
+      res <- withVisible(f(...))
       cache$set(hash, res)
-      res
+    }
+
+    if (res$visible) {
```
Comment (hadley) [inline R/memoise.r]: Could you add a little reasoning about why you added this? To me, you're obviously only calling a memoised function for its return value (not its side effects) so visibility is not very important

File: R/memoise.r (hunk 3 above)
Comment (hadley) [inline R/memoise.r]: Might also be worth adding

```R
make_call <- function(name, args) {
  stopifnot(is.name(name), is.list(args))
  as.call(c(list(name), args)
}
```

File: R/memoise.r [hunk 6]
```diff
@@ -35,64 +35,132 @@
...
+      } else {
+        invisible(res$value)
+      }
+    },
+    as.environment(list(list_call = list_call, init_call = init_call)))
+  )
+  formals(memo_f) <- f_formals
+  attr(memo_f, "memoised") <- TRUE
+
+  cache_env <- new.env(parent = envir)
```
Comment (hadley) [inline R/memoise.r]: Why should the cache inherit from the parent env?

File: cran-comments.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+* win-builder (devel and release)
+
+## R CMD check results
+There were no ERRORs, WARNINGs, or NOTEs.
+
+## Downstream dependencies
+I have also run R CMD check on 9 downstream dependencies of memoise
+(https://github.com/hadley/memoise/blob/7597e250e87322a064cd97adeea36e5bad669f7d/revdep/summary.md). All packages that I could install passed except:
+
+* lubridate: example failed most likely due to localized system (LC_TIME=de_CH.UTF-8)
```
Comment (hadley) [inline cran-comments.md]: You can add "the lubridate maintainers are aware of the problem and a currently working on a release to fix this (and other) problems"

File: R/memoise.r (hunk 1 above)
Comment (krlmlr) [inline R/memoise.r]: I think `exists()` won't work with anonymous functions.

File: R/memoise.r (hunk 2 above)
Comment (krlmlr) [inline R/memoise.r]: Now using `formals(args(f))` pattern instead -- works also with primitives.

File: R/memoise.r (hunk 4 above)
Comment (krlmlr) [inline R/memoise.r]: Not easily. `bquote()` uses NSE, I'd have to jump through hoops here.

File: R/memoise.r (hunk 6 above)
Comment (krlmlr) [inline R/memoise.r]: This was a misnomer -- renamed.

File: R/memoise.r (hunk 5 above)
Comment (krlmlr) [inline R/memoise.r]: This is #2. One possible application is a function that needs to run only once -- this function would be called for its side effects.

**Comment by krlmlr** [comment]
Thanks for the review, I have addressed all remarks. The most recent changes are https://github.com/krlmlr/memoise/pull/16.

**Comment by krlmlr** [comment]
Just updated the revdep check. Hope there won't be too many conflicts with @jimhester's pull requests.

**Comment by jimhester** [comment]
Thanks Kirill, this makes the interface much nicer!

**Pull request merged**